### PR TITLE
[flang] Generate main only when a Fortran program statement is present

### PR DESCRIPTION
**Pull request by DavidTruby**
This patch changes the behaviour for flang to only create and link to a
`main` entry point when the Fortran code has a program statement in it.

This means that flang-new can be used to link even when the program is
a mixed C/Fortran code with `main` present in C and no entry point
present in Fortran.

This also removes the `-fno-fortran-main` flag as this no longer has any functionality.

### CODE REVIEW

File: flang/lib/Optimizer/Builder/Runtime/Main.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
+//===-- Main.cpp - generate main runtime API calls ----------------*- C++
```
Comment (klausler) [inline flang/lib/Optimizer/Builder/Runtime/Main.cpp:1]: This header comment needs to be cleaned up.

**Review by klausler — APPROVED** [review]

File: flang/lib/Optimizer/Builder/CMakeLists.txt [hunk 2]
```diff
@@ -24,6 +24,7 @@ add_flang_library(FIRBuilder
   Runtime/Inquiry.cpp
   Runtime/Intrinsics.cpp
   Runtime/Numeric.cpp
+  Runtime/Main.cpp
```
Comment (kkwli) [inline flang/lib/Optimizer/Builder/CMakeLists.txt:27]: I think this list is in alphabetical order?

**Review by kkwli — APPROVED** [review]
LG

**Review by tblah — COMMENTED** [review]
Thanks for this. I'm sure it will give a big improvement to the experience of building applications with flang.

Code changes look good to me. I'm worried about how this could break existing builds using `-fno-fortran-main` and `-lFortran_main`. Do you have a transition plan for this?

**Comment by kkwli** [comment]
This change will also break backward compatibility that the old object file that contain `main` can no longer use the new compiler to link with other objects. I think we can put a warning in the release note or something like that to warn users.

**Comment by banach-space** [comment]
Great work David, thanks! Could you add some documentation explaining _where_ `main` would be coming from in the case of mixed-source compilation? In fact, is that tested anywhere?

Also, IMHO it would be good to advertise this on Discourse (thinking specifically about `-fno-fortran-main`).

**Comment by DavidTruby** [comment]
> Great work David, thanks! Could you add some documentation explaining _where_ `main` would be coming from in the case of mixed-source compilation? In fact, is that tested anywhere?
> 
> Also, IMHO it would be good to advertise this on Discourse (thinking specifically about `-fno-fortran-main`).

In mixed-source compilation where there's no `program` statement in Fortran then main should be provided by whatever other language you're linking to. In the example of C/C++ that would just mean defining `main` yourself. In the case where there's a `program` statement in the Fortran we're now defining `main` by declaring it through MLIR.

I've already posted about this on discourse, you can see the discussion here: https://discourse.llvm.org/t/rfc-link-to-fortran-main-using-embedded-linker-flags/77596. I should have put that in the PR comment!

**Comment by DavidTruby** [comment]
As regards the backwards compatibility break here, I guess my take is that this change needs to happen some time and sooner is always going to be better than later. While we are still called flang-new and marked experimental I'm inclined to think we should just rip the band aid.

We could leave `-fno-fortran-main` in as deprecated but it won't actually do anything anymore, so that might just be confusing for users.

**Review by tblah — APPROVED** [review]

**Review by pawosm-arm — APPROVED** [review]
The -fno-fortran-main flag should disappear before it finds any widespread use.

**Review by DanielCChen — APPROVED** [review]
Thanks for working on this. We have quite some Fortran C-interop test cases are affected by the "duplicate main" error.

**Pull request merged**